src/ray/util/event.cc

```diff
@@ -15,6 +15,100 @@
 #include "ray/util/event.h"
 
 namespace ray {
+///
+/// LogEventReporter
+///
+LogEventReporter::LogEventReporter(rpc::Event_SourceType source_type,
+                                   std::string &log_dir, bool force_flush)
+    : force_flush_(force_flush), log_dir_(log_dir) {
+  RAY_CHECK(log_dir_ != "");
+  if (log_dir_.back() != '/') {
+    log_dir_ += '/';
+  }
+  boost::filesystem::create_directories(log_dir_);
+  file_name_ = "event_" + Event_SourceType_Name(source_type);
+  if (source_type == rpc::Event_SourceType::Event_SourceType_CORE_WORKER ||
+      source_type == rpc::Event_SourceType::Event_SourceType_COMMON) {
+    file_name_ += "_" + std::to_string(getpid());
+  }
+  file_name_ += ".log";
+
+  log_sink_ = spdlog::rotating_logger_mt(
+      GetReporterKey() + log_dir_ + file_name_, log_dir_ + file_name_,
+      1048576 * rotate_max_file_size_, rotate_max_file_num_);
+  log_sink_->set_pattern("%v");
+}
+
+LogEventReporter::~LogEventReporter() {
+  Flush();
+  log_sink_.reset();
+}
+
+void LogEventReporter::Flush() { log_sink_->flush(); }
+
+std::string LogEventReporter::EventToString(const rpc::Event &event) {
+  std::stringstream result;
+
+  boost::property_tree::ptree pt;
+
+  auto time_stamp = event.timestamp();
+  std::stringstream time_stamp_buffer;
+  char time_buffer[30];
+  time_t epoch_time_as_time_t = time_stamp / 1000000;
+
+  struct tm *dt = localtime(&epoch_time_as_time_t);
+  strftime(time_buffer, sizeof(time_buffer), "%Y-%m-%d %H:%M:%S.", dt);
+
+  time_stamp_buffer << std::string(time_buffer) << std::setw(6) << std::setfill('0')
+                    << time_stamp % 1000000;
+
+  pt.put("time_stamp", time_stamp_buffer.str());
+  pt.put("severity", Event_Severity_Name(event.severity()));
+  pt.put("label", event.label());
+  pt.put("event_id", event.event_id());
+  pt.put("source_type", Event_SourceType_Name(event.source_type()));
+  pt.put("host_name", event.source_hostname());
+  pt.put("pid", std::to_string(event.source_pid()));
+
+  std::stringstream message_buffer;
+  const std::string &message = event.message();
+  for (int i = 0, len = message.size(); i < len; ++i) {
+    if (message[i] == '\n' || message[i] == '\r') {
+      message_buffer << "\\n";
+    } else if (message[i] == '\"') {
+      message_buffer << "\'";
+    } else {
+      message_buffer << message[i];
+    }
+  }
+
+  pt.put("message", message_buffer.str());
+
+  boost::property_tree::ptree pt_child;
+  for (auto &ele : event.custom_fields()) {
+    pt_child.put(ele.first, ele.second);
+  }
+
+  pt.add_child("custom_fields", pt_child);
+
+  std::stringstream ss;
+  boost::property_tree::json_parser::write_json(ss, pt, false);
+
+  return ss.str();
+}
+
+void LogEventReporter::Report(const rpc::Event &event) {
+  RAY_CHECK(Event_SourceType_IsValid(event.source_type()));
+  RAY_CHECK(Event_Severity_IsValid(event.severity()));
+  std::string result = EventToString(event);
+  result.pop_back();  // pop '\n'
```

Comment:
You'd better make sure the comment is full sentence starts with uppercase and ends with `.` at front of code.